core/src/main/codegen/templates/Parser.jj

```diff
@@ -5300,6 +5302,30 @@ SqlCall JsonDepthFunctionCall() :
     }
 }
 
+SqlCall JsonLengthFunctionCall() :
+{
+    final SqlNode[] args = new SqlNode[1];
+    SqlNode e;
+    final Span span;
+    List<SqlNode> behavior;
+}
+{
+    <JSON_LENGTH> { span = span(); }
+    <LPAREN> [
```

Comment:
The following code seems not to implement the syntax `JSON_LENGTH(json_doc[, path])`. Because basically the production `JsonApiCommonSyntax()` accepts input `json_doc, path`, where `JsonValueExpression(true)` accepts `json_doc`. Either of them does not suit the parameter syntax of JSON_LENGTH.
And you may want to add a test to test against the case `JSON_LENGTH(json_doc)`, to see if the parser logic is good.

Reply:
hi @zhztheplayer  Sorry for the late reply. `JSON_LENGTH(json_doc[, path])` calcite is semantically supported. I hope you can help me. I don't quite understand the implementation of `JsonApiCommonSyntax()`, but my idea is to take this `JsonApiCommonSyntax()` and tweak it to take care of the `JSON_LENGTH` syntax.  
best
qianjin

Reply:
Currently `JsonApiCommonSyntax()` only accepts parameters like `json_doc, path`, which means, the `path` expression must be specified explicitly.

I suggest to make some changes to code of `JsonApiCommonSyntax()` to make it support `json_doc[, path]`, for instance, change the current syntax code to something like:

```
SqlNode JsonApiCommonSyntax() :
{
    SqlNode e;
    List<SqlNode> args = new ArrayList<SqlNode>();
    Span span;
    SqlOperator op;
}
{
    e = JsonValueExpression(true) {
        args.add(e);
        span = Span.of(e);
    }
    ( 
        <COMMA>
        e = Expression(ExprContext.ACCEPT_NON_QUERY) {
            op = SqlStdOperatorTable.JSON_API_COMMON_SYNTAX;
            args.add(e);
        }
    |
        {
            op = SqlStdOperatorTable.JSON_API_COMMON_SYNTAX_WITHOUT_PATH;
        }
    )
    [
        <PASSING> e = JsonValueExpression(false) {
            args.add(e);
        }
        <AS> e = SimpleIdentifier() {
            args.add(e);
        }
        (
            <COMMA>
            e = JsonValueExpression(false) {
                        args.add(e);
            }
            <AS> e = SimpleIdentifier() {
                        args.add(e);
            }
        )*
    ]
    {
        return op.createCall(span.end(this), args);
    }
}
```

Then we should add a new operator `JSON_API_COMMON_SYNTAX_WITHOUT_PATH` and implement it at enumerable level.

Reply:
Sorry if i've missed a lot of context. But if JSON_LENGTH is syntactically just a regular function then I don't think we need to change Parser.jj.

Reply:
Thanks @julianhyde,
 
The implementation (also JSON_TYPE / JSON_DEPTH / JSON_PRETTY) reuses the parameter `JsonValueExpression`, which is the common JSON constructor of SQL:2016 JSON functions, By doing this the MySQL's JSON functions may also support some Non-JSON inputs, for instance, users can use `JSON_TYPE(col FORMAT AVRO)` to handle avro input (but currently we don't support avro input yet).

I personally have no objections on this[1]. Do you think we should not use `JsonValueExpression` on non-standard JSON operators?

And I think we can move `JsonValueExpression` to general `Expression` production in the fix of (CALCITE-2689)[https://issues.apache.org/jira/browse/CALCITE-2869]. Once we do that, we can also remove the MySQL JSON functions' definition from Parser.jj.

[1] https://issues.apache.org/jira/browse/CALCITE-2864?focusedCommentId=16779941&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16779941

Reply:
hi  @zhztheplayer @julianhyde Hi Thank you for your suggestion, I will continue to implement `JSON_LENGTH(json_doc[, path])` semantics by transforming `JsonApiCommonSyntax`.